### PR TITLE
[IMP] mass_mailing: owlify filter widget

### PR DESCRIPTION
**Pull request by jeh-odoo**
Owlification of the mailing filter widget

task-2984157

### CODE REVIEW

File: addons/mass_mailing/static/src/js/mailing_m2o_filter.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+ * In edit mode, it will allow to save the latest configured domain
+ * in form of the favorite filter, or to remove the store filters.
+ *
+ */
+export class FieldMany2OneMailingFilter extends Many2OneField {
+    setup() {
+        super.setup();
+        this.notification = useService("notification");
+        onMounted(this._updateFilterIcons);
+        onPatched(this._updateFilterIcons);
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mailing_m2o_filter.js:24]: "useEffect" seems to be a good candidate to apply every time you need it.

I let you grep other usages to see the syntax.

File: addons/mass_mailing/static/src/js/mailing_m2o_filter.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        // value into object, so we use 'Domain' class to convert them in same (string) format, allowing proper comparison.
+        const recordDomain = new Domain(this.props.record.data[this.props.domain_field] || []).toString();
+        const filterDomain = new Domain(this.props.record.data.mailing_filter_domain || []).toString();
+        el.classList.toggle('d-none', recordDomain === '[]');
+        const canSaveFilter = !this.props.record.data.mailing_filter_id
+            || !document.querySelector(`input#${this.props.model_field}`).value.length
+            || this.state.isFloating
+            || filterDomain !== recordDomain;
+        el.querySelector('.o_mass_mailing_save_filter_container').classList.toggle('d-none', !canSaveFilter);
+        el.querySelector('.o_mass_mailing_remove_filter').classList.toggle('d-none', canSaveFilter);
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mailing_m2o_filter.js:59]: Wondering if we could use magic "state" for this.

Like having a t-attf-style in the DOM checking a this.state.canSaveFilter and this.state.filterCount that would be magically adapted.

Could be interesting to try it out to lessen manual DOM manipulations.

File: addons/mass_mailing/static/src/js/mailing_m2o_filter.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+ * in form of the favorite filter, or to remove the store filters.
+ *
+ */
+export class FieldMany2OneMailingFilter extends Many2OneField {
+    setup() {
+        super.setup();
+        this.notification = useService("notification");
+        onMounted(this._updateFilterIcons);
+        onPatched(this._updateFilterIcons);
+        useExternalListener(document, "show.bs.dropdown", this.onFilterDropdownShown);
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mailing_m2o_filter.js:25]: That seems risky.

If any other dopdown is shown, then you will apply your behavior?

Can't we make a small Component out of this dropdown and handle events manually in there?

File: addons/mass_mailing/static/src/js/mailing_m2o_filter.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    /**
+     * On opening the drop-down, remove existing input value (if any) and set
+     * focus on input
+     *
+     * @param {event} ev
+     */
+    onFilterDropdownShown(ev) {
+        const filterInput = ev.currentTarget.querySelector('input.o_mass_mailing_filter_name');
+        filterInput.value = '';
+        setTimeout(() => filterInput.focus({ preventScroll: true }), 100);
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mailing_m2o_filter.js:73]: I tend to dislike those arbitrary setTimeout ^^

If we had a small component for the dropdown, we could use onMounted to auto-focus.

**Review by awa-odoo — COMMENTED** [review]
Hello @jeh-odoo 

Very clean port, I don't have much to say :ok_hand: 

I'll admit I did not read the tests in detail, but I bet these have been the most fun part :smile: 

Thanks & cheers!

**Comment by jeh-odoo** [comment]
Hello @awa-odoo 
Thanks for the review!
The changes are made :smile: (and the test were indeed a fun part :sweat_smile: )
Cheers!

**Comment by awa-odoo** [comment]
@robodoo r+

**Comment by tde-banana-odoo** [comment]
@robodoo retry

**Pull request closed without merging**